org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java

```diff
@@ -602,4 +599,19 @@ private static boolean isModularProject(IJavaProject project) throws JavaModelEx
 	String modName = module == null ? null : module.getElementName();
 	return modName != null && modName.length() > 0;
 }
+
+private static boolean hasSystemModule(JavaProject javaProject, IClasspathEntry classpathEntry) {
+	if (classpathEntry.getEntryKind() == IClasspathEntry.CPE_CONTAINER) {
+		IPackageFragmentRoot[] fragmentRoots = javaProject.findPackageFragmentRoots(classpathEntry);
+		if (fragmentRoots != null) {
+			for (IPackageFragmentRoot fragmentRoot : fragmentRoots) {
+				IModuleDescription module = fragmentRoot.getModuleDescription();
+				if (module != null && module.isSystemModule()) {
+					return true;
```

Comment:
@iloveeclipse see this comment: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/748#issuecomment-1430211639

This is what the build path page for libraries does. I've not determined what the compiler does , I think its done via the `instanceof JrtPackageFragmentRoot` statements in `SearchableEnvironment`.

Reply:
Simeon, I think this can be simplified. We **have** `PackageFragmentRoot` from `isOnModulePath` that is being to be added to search path, we do not want to loop over a list of **other** PFR's from the project.

Reply:
This is iterating over package fragments from the classpath **container** entry. The entry for which we are checking whether to add module infos or not.

Or am I misunderstanding you?

Reply:
isOnModulePath asks about PFR. We get a classpath entry from PFR and call `hasSystemModule`  with that entry, from which we **again** asks for **all** PFR and not that one that we've got initially. This makes no sense. We **have already** PFR we want to check, it has to be provided as argument to `hasSystemModule`.